README
iqdb - Image Query Database system
piespy@gmail.com

Distributed under the terms of the GNU General Public License,
see the file COPYING for details.

1) Compiling

- install either the libGD+libpng+libjpeg, or the ImageMagick dev package
  Using GD, on Debian: libgd2-xpm-dev libpng12-dev libjpeg62-dev
  Using ImageMagick, on Debian: libmagick9-dev, on Fedora: ImageMagick-devel)
- check the top of the Makefile for settings
- run make
- if it fails, fix the Makefile or something

2) Running

The iqdb program has two operating modes: database maintenance and query server.

The maintenance mode is to update image databases: add and remove images, set
image properties and retrieve database stats. The query server mode listens on
a TCP port for query commands and returns matching image IDs.

a) Maintenance

Two major modes of calling the maintenance mode:

$ iqdb add foo.db

This mode expects a list of image IDs and filenames on stdin, in this format:
	<ID> [<width> <height>]:<filename>

The ID is the image ID in hexadecimal. If width and height are not specified,
the values from the filename are used instead. The filename can be any image
format supported by ImageMagick; the image will be resized to 128x128 and
added to the database. Duplicate IDs are ignored. Note that iqdb does not
remember the filename associated with an image ID, you are responsible for
keeping track of that. It refers to an image exclusively by the ID.

A more complex mode allows add, removing, updating and querying images for
multiple databases at once:

$ iqdb command foo.db bar.db baz.db

This mode expects commands on stdin. For a full list see do_commands() in
iqdb.cpp. The <dbid> argument refers to the database on the command line, with
the first argument being dbid=0. Commands of particular interest are:

	add <dbid> <imgid> [<width> <height>]:<filename>
		See above.
	remove <dbid> <imgid>
		Remove the given image from the DB.
	rehash <dbid>
		This commands updates the coefficient bucket sizes.
		It is normally not needed, and is inefficient because
		it needs to read in the entire database.
		(As a special feature, `iqdb rehash foo.db' can
		be used to upgrade an old version of the DB.)
	done
		Quit and save all databases to their original file.
	quit
		Quit and save all databases to their original file.

The maintenance commands (see listen mode below) are also supported in
command mode, but perhaps not very useful here.

Changes are automatically saved in the original file upon EOF on stdin,
or when sending the "done now" or "quit now" commands.

DO NOT interrupt the program with Ctrl-C or similar until the database
is closed or it may become corrupted.


b) Server mode

In query server mode, iqdb loads the databases into memory in read-only mode
to allow the fastest image queries. No database modifications are possible.

$ iqdb listen [IP:]port [-r] [-d=<debuglevel>] [-s<IP/host>...] foo.db bar.db baz.db

Listens on the given IP:port (default localhost if no IP given) for commands,
after loading the given databases. If -r is specified and the port is
currently in use by another instance of iqdb, it will load the database and
then terminate that instance, to reduce the downtime while loading the
databases (and to leave the old server running in the rare case that it
should crash while starting up). The debug level may be set with -d.
The -s option allows limiting iqdb queries to the given source IPs. This
is useful if you cannot bind iqdb to localhost, for instance because the
query script and iqdb run on different hosts. Then you can limit iqdb
requests to the host where the script is running. Note that you should
probably also specify "-s<host-IP>" with the IP as given to the listen
argument, to allow requests from the local host, for instance to add images
to the database and to make the -r option work.

$ iqdb listen2 [IP:]port [options...] foo.db bar.db baz.db

Same as above, but listens on the given port and one port below it (i.e.
if port 5588 is specified, also listens on 5587). The lower port is higher
priority and all pending requests are serviced before the other port.

Of particular interest are the following commands:

	query <dbid> <flags> <numres> <filename>
		Find numres images most similar to given filename. If the
		filename starts with ':', use e.g. "./filename" to disambiguate
		it from the literal image data query below.
		Flags is a bitmask:
			0 = normal operation
			1 = file contains a sketch, use different weights
			2 = force grayscale match, discard color information
			8 = consider image width as set ID instead and
			    return only the best match for each set
			16= discard common coefficients (those present in at
			    least 10% of all images), this finds a
			    near-identical match much faster but the
			    non-matching images have less similarity

	query <dbid> <flags> <numres> <:size>
		As above, but if the filename argument starts with ':', the
		given number of bytes of literal image data starts on the next
		line after the command. This allows querying without first
		writing the image data to a file, and to query images that don't
		locally exist on the host running the iqdb server.

	multi_query <dbid> <flags> <numres> [+ <dbid2> <flags2> <numres2> +...] <filename>
	multi_query <dbid> <flags> <numres> [+ <dbid2> <flags2> <numres2> +...] <:size>
		Merge query results from multiple databases. This adjusts the
		individual scores to deal with varying similarity noise levels
		in different databases (the noise level depends on the number
		of images in the DB and how self-similar they are).

	query_opt <option> <arguments...>
		Set advanced options for the next query. Options are reset after
		the next query or multi_query command. Available options are:
		mask <and> <xor>
			Consider image height as bitmask and only return images
			for which ((bitmask & and) ^ xor) == 0. In other words,
			for each bit that is set in "and", the image mask's bit
			must match that of the "xor" mask. This can be used to
			limit search results to given rating masks for example.
		mindev <min.std.dev.>
			Return only results that are the given standard deviation
			above the noise level. Nothing is returned if there are
			no relevant results at all.

The "add" and "remove" commands are now supported as well, however they
only modify the memory representation of the DB and cannot be saved back
to disk later. They allow you to update the server without restarting it
but require that you also update the DB file directly.

Additionally, since version 20081123 the listen mode also supports DB
maintenance commands. In two-port mode these will only be accepted on
the lower (high-priority) port.

	load <dbid> <mode> <filename>
		Loads the given DB file into the given dbid (which must not
		be in use yet). The mode can be one of the following:
			simple	default listen mode, supports fast queries
				as well as adding/removing images
			read_only requires more memory but supports image
				ID queries (needed for duplicate finding)
			alter	default command mode, supports fast updates
				of DB files but no queries
			normal	full functionality but requires much memory
				and has slow queries, mainly useful for
				upgrading old DB versions

	drop <dbid>
		Drops the given database. With drop and load commands in the
		same connection, a database will be reloaded or replaced
		atomically, however the server will be blocked until loading
		is complete. When loading a large DB it is recommended to
		start a new copy of the server with the -r option instead.

	db_list
		Lists all loaded databases with dbid and filename.

The server has the following possible responses:

	000 iqdb ready
		Displayed whenever iqdb is ready to
		accept a command.
	100 <text>
		General informational message.
	101 <key>=<value>
		Specific parseable information.
	102 <dbid> <filename>
		Response to db_list command.
	200 <imgid> <score> <width> <height>
		Query result.
	201 <dbid> <imgid> <score> <width> <height>
		Multi-query result.
	202 <original id>=<std.dev> <dupe1 id>:<sim1> [...]
		Duplicate finder result.
	300 <text>
		General error message.
	301 <exception> <description>
		Non-fatal error message (e.g. invalid image ID, or
		unreadable image file)
	302 <exception> <description>
		Fatal error message (e.g. corrupted database)

3) Querying

The file iqdb.php holds sample PHP code to connect to a running iqdb server
and queries it for similar images.

4) Converting

Since version 20090612, iqdb will automatically detect the integer sizes
used for writing a given image database. It will automatically convert them
to its internal format on reading. Writing a non-native image database is
not possible.

This version also by default is compiled in 64-bit mode to make the image
database file platform-independent.

It is possible, but not recommended however, to have iqdb convert an image
database of an older version into this new platform-independent format. It's
better to rebuild the database from scratch. If you do want to convert the
database, first make a backup of it, in case the conversion should fail.
Then run "iqdb rehash <db-file>".

If you haven't turned off debugging, this should output the following:
	Loading db (converting data sizes)...
	Database loaded from <db-file>, has <num> images.
After loading, it will save the database again in the platform-independent
format. Pay particular attention to the number of images it reports. If
this isn't the number you expect, the conversion probably didn't work. In
that case you will have to rebuild the database from scratch using the
new version.



Changelog
iqdb-20090612:
- add imgseek mode to dupe finder script that searches for a
  given image file in the dupe database
- fix wrong transparent png background color
- add int size and endianness check
- some minor fixes to allow it to compile on Debian Lenny
- add platform-independent 64bit DB file mode and make it the default
- auto-convert on load if int sizes don't match, see README
  for details on how to do a permanent conversion to 64bit mode
- allow querying by image data literal (instead of filename)
- allow limiting iqdb queries to a few IP addresses using
  -s <IP/hostname> (can be given multiple times)
- allow to return only relevant results that exceed a minimum
  standard deviation, set using "queryopt mindev <min.std.dev.>"
- report number of results that follow for every query, using
  a "101 matches=<number>" reply
- allow setting debug level at start with -d and while running
  with the "debuglevel <level>" command

iqdb-20081123:
- make it compile with g++ 4.3.x
- add maintenance commands to list, load, save and drop databases
  without restarting the server
- listen mode can now listen on two sockets, only one of which
  supports maintenance commands, and whose pending requests will
  be serviced before those on the other (low priority) socket
- make it compile on *BSD (missing include)
- fix crash in delta_queue (ungodly memory consumption)
  on 64bit platforms

iqdb-20080718:
- support GD for image loading and resizing
  (it's the default now, edit the Makefile if you prefer ImageMagick)
- special code for efficiently resizing high-res JPEG and PNG images
  if using GD
- fixed dupe finder, call iqdb find_duplicated <dbfile> [<std.dev threshold>]
- added mkdup script to maintain filename and iqdb database and
  generate lists of files with duplicate images
- allow using custom routines for microtime_float, thumb_fname, make_thumb
  and request_srv in the PHP code, if you need to do things a little, but
  not completely, differently
- added sample code showing how to use CurlTransfer for URL
  transfers with status updates

iqdb-20080612:
- optimized delta_iterator to improve query speed by ~20%
  (now as fast or even slightly faster than before)

iqdb-20080611:
- greatly reduced memory usage (and only slightly increased query
  times) with USE_DELTA_QUEUE in the Makefile (enabled by default)
- support for returning only results matching an AND/XOR bit mask
  set the mask with the command "query_opt mask <and> <xor>"
  (mask stored internally instead of the image width)
- the rehash command in "alter" mode (e.g. when run as
  "iqdb command ...") will update bucket sizes if they are out of
  sync, to reduce memory usage
- added PHP CurlTransfer class which implements the HTTP transfer
  with regular status updates on slow or stalled transfers
- fixed a bug that caused it to ignore width and height set
  while adding an image, as opposed to with "set_res" later
- several internal code cleanups and optimizations

iqdb-20080417:
- load data needed for queries into memory in readonly/simple mode
  (can be disabled by editing the Makefile configuration)
- removed support for multiple DBs in one file
- added support for direct manipulation of file (adding/removing pics)
  for faster updates
- added support for add/remove in simple/readonly mode as well, though
  these modifications can obviously not be saved and so need to be
  done both in memory and the DB file
- the "command" and "add" options now use the new alter mode, hence
  queries are not supported (along with some other info commands)
- corrupt database data now causes iqdb to abort with exit code 10.
  this indicates that restarting it is useless until the DB is fixed.

iqdb-20080303:
- preliminary find_duplicates function
- refactored web interface into frontend and XML backend

iqdb-20080218:
- first published version


iqdb.php
<?php

$iqdb_host = "localhost";	# Host and
$iqdb_port = 5566;		# port where iqdb server is listening.
$base_dir = "/opt/iqdb/foo/";	# Directory containing image thumbnails.

# Connect to iqdb server, send it a command and parse the replies.
function request_srv($line, $err = 0) {
	global $iqdb_host,$iqdb_port;

	for ($try = 0; $try < 5; $try++) {
		@$fp = fsockopen($iqdb_host, $iqdb_port, $errno, $errstr, 30);
		if ($fp) break;

		sleep(2);
	}

	if (!$fp) {
		die("Can't connect to database.");
	}

	fwrite($fp, "$line\n");
	fflush($fp);
	$res = array("info" => array(), "values" => array(), "results" => array());

	while ($line = fgets($fp, 1024)) {
		list($code,$rest) = split(" ", chop($line), 2);
		switch($code) {
			case 100: array_push($res["info"], $rest); break;
			case 101: list($var,$val) = split("=", $rest, 2); $res["values"][$var] += $val; break;
			case 200: array_push($res["results"], "-1 $rest"); break;
			case 201: array_push($res["results"], $rest); break;
			case 300:
			case 301:
			case 302: $res["err"] = $rest; if ($err) die("Error: $rest"); break;
			default: die("Unsupported result code $code $rest");
		}
	}
	fclose($fp);

	return $res;
}

# Ask iqdb server for images most similar to given file.
function request_match($file) {
	global $base_dir;

	$res = request_srv("count 0\nquery 0 0 16 $file\ndone now\n");
	$result = array("values" => $res["values"], "match" => array());

	foreach ($res["results"] as $line) {
		list($dbid,$id,$sim,$width,$height) = split(" ", $line);

		# Resolve imgid => md5
		$pref = substr($id,0,1)."/".substr($id,1,1)."/".substr($id,2,1);
		$files = glob("$base_dir/$pref/$id*");
		if (count($files) > 1) continue;
		if (count($files) == 0) $files = glob("$base_dir/xx/$id-*");
		if (count($files) > 1) continue;
		if (!ereg("/([0-9a-fx\/]+/([0-9a-f-]+)\.jpg)$", $files[0], $eres)) continue;
		array_push($result["match"], array('sim' => $sim, 'md5' => substr($eres[2],-32),
						'fname' => $eres[1], 'width' => $width, 'height' => $height));
	}
	if (!count($result["match"])) return array('err' => $res["err"]);
	return $result;
}

$res = request_match("test.jpg");
var_dump($res);

?>
